src/detect-ikev1-chosen-sa.c

```diff
@@ -0,0 +1,286 @@
+    sigmatch_table[DETECT_AL_IKEV1_CHOSEN_SA].AppLayerTxMatch = DetectIkev1ChosenSaMatch;
+    sigmatch_table[DETECT_AL_IKEV1_CHOSEN_SA].Setup = DetectIkev1ChosenSaSetup;
+    sigmatch_table[DETECT_AL_IKEV1_CHOSEN_SA].Free = DetectIkev1ChosenSaFree;
+    sigmatch_table[DETECT_AL_IKEV1_CHOSEN_SA].RegisterTests = IKEV1ChosenSaRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+
+    DetectAppLayerInspectEngineRegister("ikev1.chosen_sa_attribute",
+            ALPROTO_IKEV1, SIG_FLAG_TOCLIENT, 2,
+            DetectEngineInspectIkev1ChosenSaGeneric);
+
+    g_ikev1_chosen_sa_buffer_id = DetectBufferTypeGetByName("ikev1.chosen_sa_attribute");
+}
+
+static int DetectEngineInspectIkev1ChosenSaGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
+                                          f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \internal
+ * \brief Function to match SA attributes of a IKEv1 state
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the Ikev1 Transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into DetectIkev1ChosenSaData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectIkev1ChosenSaMatch(DetectEngineThreadCtx *det_ctx,
+                                   Flow *f, uint8_t flags, void *state,
+                                   void *txv, const Signature *s,
+                                   const SigMatchCtx *ctx)
+{
+    SCEnter();
+
+    const DetectIkev1ChosenSaData *dd = (const DetectIkev1ChosenSaData *)ctx;
+
+    uint32_t value;
+    if (!rs_ikev1_state_get_sa_attribute(txv, dd->sa_type, &value))
+        SCReturnInt(0);
+    if (value == dd->sa_value)
+        SCReturnInt(1);
+    SCReturnInt(0);
+}
+
+/**
+ * \internal
+ * \brief Function to parse options passed via ikev1.chosen_sa_attribute keywords.
+ *
+ * \param rawstr Pointer to the user provided options.
+ *
+ * \retval dd pointer to DetectIkev1ChosenSaData on success.
+ * \retval NULL on failure.
+ */
+static DetectIkev1ChosenSaData *DetectIkev1ChosenSaParse (const char *rawstr)
+{
+    /*
+     * idea: do not implement one c file per type, invent an own syntax:
+     * ikev1.chosen_sa_attribute:"encryption_algorithm=4"
+     * ikev1.chosen_sa_attribute:"hash_algorithm=8"
+     */
+    DetectIkev1ChosenSaData *dd = NULL;
+    int ret = 0, res = 0;
+    int ov[MAX_SUBSTRINGS];
+    char attribute[100];
+    char value[100];
+
+    ret = DetectParsePcreExec(&parse_regex, rawstr, 0, 0, ov, MAX_SUBSTRINGS);
```

Comment:
Could we do this in Rust ?

Reply:
just wanted to be consistent with the existing keyword implementations.

Reply:
We are now trying to move more code to Rust cf https://redmine.openinfosecfoundation.org/issues/3195

One example of signature code parsing in Rust is given :
https://github.com/OISF/suricata/blob/master/rust/src/dns/detect.rs#L34